runtime/src/bank/fee_distribution.rs

```diff
@@ -0,0 +1,893 @@
+}
+
+#[derive(Error, Debug, PartialEq)]
+enum DepositFeeError {
+    #[error("fee account became rent paying")]
+    InvalidRentPayingAccount,
+    #[error("lamport overflow")]
+    LamportOverflow,
+    #[error("invalid fee account owner")]
+    InvalidAccountOwner,
+}
+
+impl Bank {
+    // Distribute collected transaction fees for this slot to collector_id (= current leader).
+    //
+    // Each validator is incentivized to process more transactions to earn more transaction fees.
+    // Transaction fees are rewarded for the computing resource utilization cost, directly
+    // proportional to their actual processing power.
+    //
+    // collector_id is rotated according to stake-weighted leader schedule. So the opportunity of
+    // earning transaction fees are fairly distributed by stake. And missing the opportunity
+    // (not producing a block as a leader) earns nothing. So, being online is incentivized as a
+    // form of transaction fees as well.
+    //
+    // On the other hand, rent fees are distributed under slightly different philosophy, while
+    // still being stake-weighted.
+    // Ref: distribute_rent_to_validators
+    pub(super) fn distribute_transaction_fees(&self) {
+        let collector_fees = self.collector_fees.load(Relaxed);
+        if collector_fees != 0 {
+            let (deposit, mut burn) = self.fee_rate_governor.burn(collector_fees);
+            if deposit > 0 {
+                let validate_fee_collector = self.validate_fee_collector_account();
+                match self.deposit_fees(
+                    &self.collector_id,
+                    deposit,
+                    DepositFeeOptions {
+                        check_account_owner: validate_fee_collector,
+                        check_rent_paying: validate_fee_collector,
+                    },
+                ) {
+                    Ok(post_balance) => {
+                        self.rewards.write().unwrap().push((
+                            self.collector_id,
+                            RewardInfo {
+                                reward_type: RewardType::Fee,
+                                lamports: deposit as i64,
+                                post_balance,
+                                commission: None,
+                            },
+                        ));
+                    }
+                    Err(err) => {
+                        // notify validator that their fees were burned
+                        error!(
+                            "Burned {deposit} lamport transaction fee instead of sending to {} due to {err}",
+                            self.collector_id
+                        );
+                        burn += deposit;
+                    }
+                }
+            }
+            self.capitalization.fetch_sub(burn, Relaxed);
+        }
+    }
+
+    // Deposits fees into a specified account and if successful, returns the new balance of that account
+    fn deposit_fees(
```

Comment:
We need the same checks (including rent state transition) in both cases, that's why I created this function. If we collect transaction fees into an account whose post balance is not greater than or equal to the rent exempt balance, we need to skip the fee deposit because otherwise we allow creating rent paying accounts. Make sense?

Reply:
Yeah. 
It seems that the current code still allows transaction fees to be collected into an account whose post balance is not greater than or equal to the rent exempt balance. Should we create a separate PR to fix this for current code and backport it to v1.17?

Reply:
I'm not sure what you mean, can you elaborate more? By the "current code" you mean the code in this PR right? And under what conditions are you saying that the code "still allows transaction fees to be collected into an account whose post balance is not greater than or equal to the rent exempt balance"?

Reply:
By "current code", I mean master. 

In master, we don't validate the post balance of the account that transaction fees are deposit to. If the balance of the account pre-transaction fee was zero, then after the transaction fee deposit, we introduce a "new" rent paying account, right?

The new code (i.e. deposit_fee fn) in the PR will prevent this from happening. What I mean is that we pull this piece out in a new PR to patch master separately.

Reply:
I see, I'd rather not break up apart this PR unless we have a good reason to. I purposely setup this PR to introduce the shared `deposit_fee` function to ensure that all invalid cases are handled for each type of fee collection. Before, without a common function, we missed the rent paying edge case for collected transaction fees.

Reply:
OK. I think we might want to port this feature back to 1.17 to close those edge cases as soon as possible.